RocketChat/Rocket.Chat
# [NEW] Livechat voip/queue events

## Proposed changes (including videos or screenshots)
This PR implements continuous monitor in the asterisk connector. Asterisk continuously generates a stream of events on various activities. Management server user defined in asterisk's manager.conf determines the events to be sent on this user. This PR's main focus is to find out calls in the queue and source queue of a call.

To do this, it monitors following events
    

    agentcalled (For finding out the source queue of a call)  (message is sent to a to a single agent)
    queuecallerjoin   (for calls in the queue) (message is sent to all the extensions in the queue)
    agentconnect (for calls in the queue) (message is sent to all the extensions in the queue)
    queuememberadded (for calls in the queue) (message is sent to all the extensions in the queue)
    queuememberremoved (for calls in the queue) (message is sent to all the extensions in the queue)
    queuecallerabandon (for calls in the queue) (message is sent to all the extensions in the queue)
 
These events are then captured, processed and necessary information extracted. In some cases, the code also uses other connector commands (e.g queue_status) to determine certain parameters which are necessary to derive certain information.

e.g When a customer calls a queue, we need to find out all the members of the queue. This is achieved using queue_status command. Once the members are found necessary information is sent to all the members in the queue.

When the members receive the notification, the have aggregator, which makes some calculation and determines the calls waiting in a queue. 


## Issue(s)
There are some improvements necessary in the future :
1. We do not have a way to configure the events we want to watch for from our admin page. Currently these values are hardcoded. This should be configurable based on feature (source queue of a call, calls waiting in queue). So to support a feature, the continuous monitor uses certain events.
2. There is a lot of data that asterisk sends on the connection. This can be configured by correctly configuring the asterisk user on manager.conf of asterisk. This needs to be documented.
3. We do not have a strategy to reinitiate the connection or check the connectivity.
## Steps to test or reproduce


## Further comments
Testing Notes :
For determining source queue of a call :
1. Verify that the user has correct information when the call is ringing. Also verify that the user whose call is ringing gets this information. 
2. When user is a part of 2 queues, verify that user receives correct information.
3. When user does not accept the call, verify that the information is cleared.

For determining calls waiting in the queue :
1. Verify that user receives the aggregated calls in queue information when the agent is part of multiple queues.
2. Verify that on customer calling queue, agent accepting call, agent added/removed from the queue, customer abandoning queue events are properly managed so that the agent has correct and the latest view.
